### PR TITLE
Fix issue with negative indexes and empty strings

### LINKED ISSUE #9: 'String.charAt' infinitely loops on empty strings and negative indices
```sml
String.charAt("", ~1)
```

This will cause a hang, since you'll recursively call `charAt("", 0 + ~1)` which is just `charAt("", ~1)`.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #9.

### CODE REVIEW

File: ponyo/Ponyo_String_/Ponyo_String.sml [hunk 1]
```diff
@@ -15,11 +15,16 @@ struct
...
     and charAt (source: string, index: int) : char =
-        if index < 0 then charAt (source, length source + index)
-        else if index > length (source)
-	    then raise IndexError (source, index)
-	else
-	    List.nth (explode source, index)
+        let
+            val actualIndex =
+                if index < 0 then
+                    length(source) + index
```
Comment (eatonphil) [inline ponyo/Ponyo_String_/Ponyo_String.sml]: In the original there was a recursive step here.

Comment (DanielRosenwasser) [inline ponyo/Ponyo_String_/Ponyo_String.sml]: I originally rewrote it that way, but as you can see in 4ccf458e34ece5beae8f4e8a7eb3c9ea34ffd9f9, it becomes cleaner this way.

Comment (DanielRosenwasser) [inline ponyo/Ponyo_String_/Ponyo_String.sml]: You also lose context on the original index when erroring if you use the recursive step

Comment (eatonphil) [inline ponyo/Ponyo_String_/Ponyo_String.sml]: The recursion is so that you can specify a negative number whose absolute value is greater than the length of the string. If you disagree with that, I think that should be a separate PR or issue. Unless I missed it, I think this won't allow that.

Comment (DanielRosenwasser) [inline ponyo/Ponyo_String_/Ponyo_String.sml]: It won't, but I don't necessarily see why that's super useful. I'll remove the last commit from this review.
